### PR TITLE
Added ECE mapping for DIII-D

### PR DESCRIPTION
**Pull request by AreWeDreaming**
[WIP] Not tested. Requires unit test.

### CODE REVIEW

**Comment by AreWeDreaming** [comment]
Fixed all the bugs and successfully tested for DIII-D shot 185069.

**Comment by AreWeDreaming** [comment]
I think this should be ready to merge.

File: omas/machine_mappings/d3d.py [hunk 1]
```diff
@@ -462,6 +462,75 @@ def thomson_scattering_data(ods, pulse=133221, revision='BLESSED', _measurements
...
 
+def electrpn_cyclotron_emission_hardware(ods, pulse=133221, fast=False):
+    """
+    Gathers DIII-D Electron cyclotron emission locations
+
+    :param pulse: int
+
+    :param revision: string
+        Thomson scattering data revision, like 'BLESSED', 'REVISIONS.REVISION00', etc.
+    """
```
Comment (orso82) [inline omas/machine_mappings/d3d.py:473]:
```suggestion
    """
    Gathers DIII-D Electron cyclotron emission locations

    :param pulse: int

    :param fast: bool
        Use data sampled at high frequency
    """
```

File: omas/machine_mappings/d3d.py [hunk 2]
```diff
@@ -462,6 +462,75 @@ def thomson_scattering_data(ods, pulse=133221, revision='BLESSED', _measurements
...
+@machine_mapping_function(__all__)
+def electron_cyclotron_emission_data(ods, pulse=133221, _measurements=True, fast=False):
+    """
+    Loads DIII-D Thomson measurement data
+
+    :param pulse: int
+
+    :param revision: string
+        Thomson scattering data revision, like 'BLESSED', 'REVISIONS.REVISION00', etc.
+    """
```
Comment (orso82) [inline omas/machine_mappings/d3d.py:486]:
```suggestion
    """
   Loads DIII-D Electron cyclotron emission data

    :param pulse: int

    :param fast: bool
            Use data sampled at high frequency
    """
```

File: omas/machine_mappings/d3d.py [hunk 3]
```diff
@@ -462,6 +462,75 @@ def thomson_scattering_data(ods, pulse=133221, revision='BLESSED', _measurements
...
+    ece_map = mdsvalue('d3d', treename='ELECTRONS', pulse=pulse, TDI=query).raw()
+    N_time = len(ece_map['TIME'])
+    N_ch = ece_map['NUMCH'].item()
+    if _measurements:
+        query = {}
+        for ich in range(1,N_ch+1):
+            query[f'T{ich}'] = TECE + '{0:02d}'.format(ich)
+        query['TIME'] = f"dim_of({TECE + '01'})"
+    ece_data = mdsvalue('d3d', treename='ELECTRONS', pulse=pulse, TDI=query).raw()
+    # Read the Thomson scattering hardware map to figure out which lens each chord looks through
```
Comment (orso82) [inline omas/machine_mappings/d3d.py:507]:
```suggestion
```

**Review by orso82 — COMMENTED** [review]
Thanks @AreWeDreaming 
I have few minor comments

**Comment by AreWeDreaming** [comment]
Sorry for my tardiness, just got back from vacation. I committed all of your suggestions. Ready to merge?

**Comment by AreWeDreaming** [comment]
I'll take a look at that test failure.

**Comment by AreWeDreaming** [comment]
Ok, so I spent too much time on this already. The problem is that it is very hard to debug the failing test due to the `exec` statement in line `316` in `omas_machine.py`. There is an error somewhere in `thomson_scattering_hardware` in d3d.py. However, it seems that I cannot directly import `d3d.py` as a module to gain access to the functions in that module. 
Should I merge all the commits made to master into this branch? I did this locally on my machine and that did not help.

**Comment by orso82** [comment]
@AreWeDreaming I'll look into it

**Comment by AreWeDreaming** [comment]
@orso82 If you have the time I would not mind looking at it together. Let me know on discord if you are interested.

**Comment by orso82** [comment]
@AreWeDreaming I finally found time to go over your PR, made some changes, and merged. Thank you for your contribution!

**Pull request merged**